ambry-api/src/main/java/com.github.ambry/router/CryptoServiceFactory.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright 2017 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.router;
+
+/**
+ * CryptoService Factory to assist in fetching an instance of {@link CryptoService} for a given key
+ * T refers to the key type that this {@link CryptoServiceFactory} is compatible with.
+ * Ensure that {@link KeyManagementService} implementation is compatible with the same key type.
+ */
+public interface CryptoServiceFactory<T> {
+
+  /**
+   * Instantiates and returns the {@link CryptoService} for a given key.
+   * @return the {@link CryptoService} instantiated for the given key
+   */
+  CryptoService<T> getCryptoService() throws InstantiationException;
```

Comment:
`InstantiationException` is a bit annoying to use since it doesn't let you supply a cause exception to its constructor. Perhaps these factory methods can just throw `Exception`.